libc: common: add support for iso c11 threads

cfriedt opened the pull request:
**Note: Compliance issues are false positives**

Capitalize on newly added support for dynamic thread stacks and the existing pthread support to provide an implementation of ISO C11 `<threads.h>`.

For more information about the ISO C standard, please see
https://www.iso.org/standard/74528.html

Commit:
libc: common: add support for iso c11 threads

This change capitalizes on newly added support for dynamic
thread stacks and the existing pthread support to provide
an implementation of the ISO C11 `<threads.h>` API.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
tests: libc: add tests for iso c11 threads

Add tests to verify functionality of the C11 `<threads.h>` API.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
<img width="624" alt="Screenshot 2023-07-24 at 9 51 17 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/63c930fa-8e41-451a-8256-0d097271f2e0">

cfriedt commented on lib/libc/minimal/source/thrd/thrd.c:
```diff
@@ -0,0 +1,84 @@
+
+void thrd_yield(void)
+{
+	k_yield();
```
For consistent layering, this should be replaced with `pthread_yield()`, but it's effectively the same thing.

ycsin replied:
how about [`sched_yield`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/posix/sched.h#L40)?

keith-packard replied:
I think cfriedt wanted to ensure that the whole c11 thread API was implemented in terms of standard pthreads APIs; poking through to `k_yield` exposes a gap in that layer.

cfriedt replied:
I did add https://github.com/zephyrproject-rtos/zephyr/issues/60757 but then closed it shortly after ycsin told me it was deprecated, and it's no longer on the opengroup site.

For reference
https://man7.org/linux/man-pages/man3/pthread_yield.3.html

I guess I don't have anything against adding `pthread_yield()` even if it is deprecated, but technically `sched_yield()` would work too.

cfriedt replied:
Fixed!

cfriedt commented on tests/lib/c_lib/prj.conf:
```diff
@@ -4,3 +4,7 @@ CONFIG_TEST_USERSPACE=y
 CONFIG_MINIMAL_LIBC_NON_REENTRANT_FUNCTIONS=y
 CONFIG_MINIMAL_LIBC_RAND=y
+
+CONFIG_POSIX_API=y
```
We'll likely need some additional Kconfig, dependencies, etc.

cfriedt commented:
Just looking for some early feedback

keith-packard reviewed (commented):
This is looking fine. Is there any reason to depend upon the C library thread.h? Or could Zephyr provide its own version of that file directly?

cfriedt commented:
> This is looking fine. Is there any reason to depend upon the C library thread.h? Or could Zephyr provide its own version of that file directly?

No reason to explicitly not make it common in Zephyr that I can think of - @keith-packard, were you thinking it might be better to put this into `lib/libc/common/`?

keith-packard commented:
> No reason to explicitly not make it common in Zephyr that I can think of - @keith-packard, were you thinking it might be better to put this into `lib/libc/common/`?

It's almost something more common than libc/common -- shouldn't we always use this code, even if the C library provides alternatives?

cfriedt commented:
Need #60964 and #60961 to be merged before this.

Commit:
lib: libc: common: add C11 mutex implementation

Add support for C11 mutexes to go with C11 threads.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
tests: lib: c_lib: add tests for C11 mutexes

Add tests to cover C11 mutexes

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
Fixed another regression in `pthread_mutex_timedlock()` (moved to a separate PR).

cfriedt commented on tests/lib/c_lib/src/test_thrd.c:
```diff
@@ -0,0 +1,161 @@
+
+	do {
+		k_msleep(100);
+	} while (!detached_thread_is_probably_done);
```
Fixed!

cfriedt commented:
* added support for C11 mutexes

Commit:
libc: common: support for C11 condition variables

Add C11 condition variable support to go with C11 threads
and mutexes.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
tests: lib: c_lib: tests for C11 condition variables

Add tests for C11 condition variables.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
* filled-in `cnd` testsuite

Commit:
libc: common: support for C11 thread-specific storage

Add C11 thread-specific storage (tss) support to go with C11 threads
and mutexes.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
tests: lib: c_lib: tests for C11 thread-specific storage

Add tests for C11 thread-specific storage.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
libc: common: support for C11 call_once()

Add C11 call_once() support to go with C11 threads
and mutexes.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
tests: lib: c_lib: tests for C11 call_once()

Add tests for C11 call_once().

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
* added thread-specific storage (tss) and `call_once()` support

cfriedt commented:
Maybe I'm missing something, but it seems that there are a number of "Testsuite mismatch" errors.

Previously, it was possible to have multiple `ZTEST_SUITE()` declarations per actual testsuite. This is kind of nice, because even though some units are related as part of the same library, the tests may require different `setup()`, `teardown()`, `before()`, or `after()` functions, as well as different test `fixture`s.

```shell
INFO    - 1) tests/lib/c_lib/libraries.libc on mps2_an385 failed (Testsuite mismatch)
INFO    - 2) tests/lib/c_lib/libraries.libc.picolibc on mps2_an385 failed (Testsuite mismatch)
INFO    - 3) tests/lib/c_lib/libraries.libc.newlib on mps2_an385 failed (Testsuite mismatch)
INFO    - 4) tests/lib/c_lib/libraries.libc.newlib_nano on mps2_an385 failed (Testsuite mismatch)
INFO    - 5) tests/lib/c_lib/libraries.libc.minimal.strerror_table on mps2_an385 failed (Testsuite mismatch)
INFO    - 6) tests/lib/c_lib/libraries.libc.minimal.no_strerror_table on mps2_an385 failed (Testsuite mismatch)
```

In any cases, each of the testsuites above runs successfully with
```shell
west build -p auto -b mps2_an385 -t run -T <suite>
```

cfriedt commented:
@nashif - it seems to me that there are some false negatives reported by twister. Can you please take a look?

https://github.com/zephyrproject-rtos/zephyr/actions/runs/5976216865/job/16213675832?pr=60759

nashif commented on include/threads.h:
```diff
@@ -0,0 +1,102 @@
+/*
```
inldude/threads.h? hmmm....

cfriedt replied:
i tried moving it to `libc/common/include` but then picolibc gave me an error.

nashif replied:
IMO this needs some discussion and broader review, putting this at this level is just a source for problems and confusion.

nashif commented:
> @nashif - it seems to me that there are some false negatives reported by twister. All of these testsuites are passing FWICT, but there is a test suite mismatch (which I think is incorrect) that is getting reported
> 
> Can you please take a look?

I think I know what the issue is.. looking for a solution.

nashif commented:
> Can you please take a look?

so, this is because we are not parsing the source files correctly, twister does not do recursive globbing and things in sub-dirs are not being scanned. Please move things like `src/thrd/cnd.c` to something line `src/thrd_cnd.c` for now. Will look into adding support for recursive scanning later on.

cfriedt commented:
> @cfriedt so, this is because we are not parsing the source files correctly

@nashif - fixed in #61933

cfriedt commented:
* rebased

Commit:
tests: lib: c_lib: separate libc tests into smaller suites

C library testing is mainly there to support what is
necessary to support Zephyr. We do test a variety of libcs
currently, which is where YAML comes in handy.

However, the main libc testsuite can be overkill for testing
some things, and might not be suitable for testing optional
features.

Create a 'common' subdirectory for common libc tests.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

cfriedt commented:
* separated libc tests into common, thrd

dkalowsk commented on lib/libc/common/source/thrd/mtx.c:
```diff
@@ -0,0 +1,99 @@
+		return thrd_error;
+	}
+
+	switch (pthread_mutex_init(mutex, attrp)) {
```
for 2 results, is it really worth having a switch or just a simple if/else?  I won't repeat the question on the other places where this happens as well.

Not really a requesting a change, just a minor question on code intent.  A `switch` implies that more conditional results will be of value in the future, but any of the other possible results for this don't matter, they are just failures.

cfriedt replied:
> A `switch` implies that more conditional results will be of value in the future, but any of the other possible results for this don't matter, they are just failures.

Yes, that's exactly it. I wanted to structure it so that errors could easily be more specialized or be corrected with ease.

Since you were curious though, here is a comparison 
https://godbolt.org/z/GzPv4nf5c

with default optimizations
```diff
diff -Nua /tmp/*.S
--- /tmp/with-if-stmt.S	2023-09-07 19:52:25
+++ /tmp/with-switch-stmt.S	2023-09-07 19:52:44
@@ -10,11 +10,11 @@
         ldr     x0, [sp, 40]
         bl      pthread_mutex_init
         cmp     w0, 0
-        beq     .L2
-        mov     w0, 2
+        bne     .L2
+        mov     w0, 0
         b       .L3
 .L2:
-        mov     w0, 0
+        mov     w0, 2
 .L3:
         ldp     x29, x30, [sp], 48
         ret

```

with `-O3`
```diff
<intentionally empty>
```

dkalowsk reviewed (dismissed):
A relatively minor question on style preference that shouldn't slow down merging.  After having done multiple re-reviews, I don't see anything on this that appears obviously wrong.  Nice job!

nashif reviewed (dismissed):
location of new headers needs some more discussion.

cfriedt commented:
Arch Meeting: 2023-09-12: not enough time to cover this topic.

cfriedt commented:
I moved the `threads.h` header file to `lib/libc/common/include`.

@nashif - does that address your concerns?

The PicoLibC error I encountered is fixed.

Presumably due to include path priority, PicoLibC was including it's own [threads.h](newlib/libc/include/threads.h), but that version included `<machine/_threads.h>` which does not exist in Zephyr's fork. So I added that file too.

However, I also wanted to go a step further and ensure that this would still be usable with external toolchains, such as GNUARM Embedded. Following the instructions [here](https://lindevs.com/install-arm-gnu-toolchain-on-ubuntu), as well as the usual [GSG](https://docs.zephyrproject.org/latest/develop/beyond-GSG.html#install-a-toolchain), the testsuite passes.
```shell
export ZEPHYR_TOOLCHAIN_VARIANT=cross-compile
export CROSS_COMPILE=/opt/gcc-arm-none-eabi/bin/arm-none-eabi-
west build -p auto -b mps2_an385 -t run tests/lib/c_lib/thrd
```

Also succeeds with LLVM
```shell
export ZEPHYR_TOOLCHAIN_VARIANT=llvm
west build -p auto -b qemu_x86 -t run tests/lib/c_lib/thrd
```

cfriedt commented:
![Screenshot 2023-09-16 at 3 53 50 PM](https://github.com/zephyrproject-rtos/zephyr/assets/983494/1f6f472d-3b07-427c-9495-5f1f62c14767)

keith-packard commented:
> Presumably due to include path priority, PicoLibC was including it's own [threads.h](newlib/libc/include/threads.h), but that version included `<machine/_threads.h>` which does not exist in Zephyr's fork. So I added that file too.

Hrm. We should make sure picolibc's header files all come after anything in Zephyr so that Zephyr can replace anything it needs to. Fixing things in picolibc is definitely an option if necessary, of course, but there is a lag between getting that fixed and having it land in an SDK that people are using.

cfriedt commented:
> > Presumably due to include path priority, PicoLibC was including it's own [threads.h](newlib/libc/include/threads.h), but that version included `<machine/_threads.h>` which does not exist in Zephyr's fork. So I added that file too.
> 
> Hrm. We should make sure picolibc's header files all come after anything in Zephyr so that Zephyr can replace anything it needs to. Fixing things in picolibc is definitely an option if necessary, of course, but there is a lag between getting that fixed and having it land in an SDK that people are using.

It's a relatively minor change if we ever needed to shuffle contents from `machine/_threads.h` to `threads.h` in zephyr's repo but if the current layout makes it easier to integrate other newlib-based libcs, that is fine with me too 🤷‍♂️

nashif commented:
> @nashif - does that address your concerns?

Yes, but it would be good to give an overview of this, plans and get more feedback.

dkalowsk reviewed (dismissed):
took longer to re-review, but done... thanks for the effort on this!

cfriedt commented:
> > @nashif - does that address your concerns?
> 
> Yes, but it would be good to give an overview of this, plans and get more feedback.

I've been attending the Arch Review meeting for weeks but this has not been on the agenda. Is there a reason it keeps getting pushed back?

aescolar commented on tests/lib/c_lib/thrd/testcase.yaml:
```diff
@@ -0,0 +1,29 @@
+    - mps2_an385
+tests:
+  libraries.libc.c11_threads.minimal:
+    tags: minimal_libc
```
Nowadays this will default to picolibc, so this would need to specifically select CONFIG_MINIMAL_LIBC

cfriedt replied:
Heh.. shows how long this one has been in development 🤣

aescolar commented on tests/lib/c_lib/thrd/prj.conf:
```diff
@@ -4,3 +4,9 @@ CONFIG_TEST_USERSPACE=y
 CONFIG_ZTEST_FATAL_HOOK=y
 CONFIG_MINIMAL_LIBC_NON_REENTRANT_FUNCTIONS=y
```
Nit: maybe it would be a bit nicer if these 2 where in the testcase configuration for minimal libc or an overlay?

cfriedt replied:
Maybe even in the testcase.yaml. Good suggestion.

galak commented on lib/libc/common/Kconfig:
```diff
@@ -66,3 +66,11 @@ config COMMON_LIBC_STRNLEN
+config COMMON_LIBC_THRD
+	bool "C11 <threads.h> API support"
+	depends on DYNAMIC_THREAD
+	depends on POSIX_API
```
Worth add a comment here that the POSIX_API `depends` is only because its implemented utilizing that and could be re-implemented using native zephyr api's

carlescufi commented:
Architecture WG:

- @cfriedt explains that now that we have dynamic thread stacks it is possible to use the C standard to create threads
- @cfriedt explains that this is supposed to work with any of the C standard libraries that we support
- This is pulling in the POSIX API (because it's implemented on top of pthreads), but this can be optimized out later in the cycle. This is not trivial to implement, because the POSIX layer is already doing a lot of what's necessary to implement the API
- @carlescufi: the path of `threads.h` is a bit unusual. Why isn't it in `minimal/include` instead of `common/include`? Why not rely on the standard library-provided headers like we do for the rest? @cfriedt states that the reason is that threads are a new feature and the header is not included in many of those libraries
- @cfriedt states that, in the future, this will support multiple "backends". At that point the infrastructure can be shared between `threads.h` and POSIX so that `threads.h` can be implemented on top of the native Zephyr thread APIs using the common infrastructure.

cfriedt commented:
@aescolar, @nashif - comments addressed (updated test filters, added Kconfig comment)
@dkalowsk - would you re-ack?

ycsin commented on tests/lib/c_lib/strerror/prj.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_ZTEST=y
+CONFIG_ZTEST_NEW_API=y
```
```suggestion
```

`CONFIG_ZTEST_NEW_API` has been removed #63723

cfriedt replied:
Fixed!

ycsin commented on tests/lib/c_lib/thrd/prj.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_ZTEST=y
+CONFIG_ZTEST_NEW_API=y
```
```suggestion
```

cfriedt replied:
Fixed!

cfriedt commented:
@aescolar - just added an `arch_exclude` for `posix` to skip the one test that was failing.

https://github.com/zephyrproject-rtos/zephyr/actions/runs/6653568254/job/18079862854?pr=60759

A bit of a lofty goal, but I would really like to try to be able to exercise Zephyr's POSIX API on native posix by LTSv3. I hope we can work together to make that a reality.

@dkalowsk, @ycsin, @nashif - should be good to approve once this is green.

aescolar commented:
> A bit of a lofty goal, but I would really like to try to be able to exercise Zephyr's POSIX API on native posix by LTSv3. I hope we can work together to make that a reality.

You can, you just need select picolibc

cfriedt commented:
> > A bit of a lofty goal, but I would really like to try to be able to exercise Zephyr's POSIX API on native posix by LTSv3. I hope we can work together to make that a reality.
> 
> You can, on native_sim. You just need to select picolibc

Ideally, it would work for others too - e.g. minimal, newlib. Maybe not the niche / target-specific libcs.

aescolar commented:
> Ideally, it would work for others too - e.g. minimal, newlib. Maybe not the niche / target-specific libcs.

With minimal libc it will work also (subject to whatever minimal libc supports). newlib cannot be used with native_sim as that c library is provided by the toolchain.

cfriedt commented:
* rebased after #63941 to resolve merge conflicts. That was my fault ;-)

cfriedt commented:
@keith-packard - any thoughts?

cfriedt commented:
@dkalowsk - rebased and all green, if you would like to approve again.

aescolar reviewed (dismissed):
The -1 is due to the test exclusion on the posix arch which should not be.
The others are mostly nits.

cfriedt commented:
* made changes requested by @aescolar

aescolar commented on tests/lib/c_lib/thrd/testcase.yaml:
```diff
@@ -0,0 +1,40 @@
+    tags: minimal_libc
+    filter: CONFIG_MINIMAL_LIBC_SUPPORTED
+    platform_exclude:
+      - native_posix
```
The idea was to put this in the common part of the tests definition

cfriedt replied:
Fixed!

cfriedt commented:
@aescolar, @ycsin, @dkalowsk - it's green. hopefully that works

cfriedt commented:
* Rebased again to resolve conflicts
* added module and notls permutations for picolibc to the new testsuites
* added extra configs to provide `rand()` and `srand()` for minimal libc in common testsuite
* reorder test configurations alphabetically
* re-ran entire testsuite locally

<img width="773" alt="Screenshot 2023-11-08 at 12 20 27 PM" src="https://github.com/zephyrproject-rtos/zephyr/assets/983494/1d21ed2a-ca68-44f1-a41d-67817e6ef9bd">

Sorry to ask for re-approvals again, @aescolar, @jhedberg , @ycsin , @dkalowsk.

If you have a moment @keith-packard, please take a look.

It's probably safe to merge given there have been no code changes for ~4 months (just rebasing to fix merge conflicts and updating testcase.yaml over and over).

jgl-meta reviewed (dismissed):
lgtm!

dkalowsk reviewed (dismissed):
One minor note I saw on this review, but honestly it can wait for a future code clean up.  Looks good to me!

cfriedt commented on include/zephyr/posix/sys/stat.h:
```diff
@@ -114,7 +114,7 @@ struct stat {
 #define st_ctime st_ctim.tv_sec
 #if defined(__linux) && defined(__x86_64__)
-	__uint64_t __glibc_reserved[3];
+	uint64_t __glibc_reserved[3];
```
Yeah, that was a new one - it's because we now support running on `native_sim` (and this is pulled in via picolibc / pthreads headers).

The Great POSIX Header Cleanup (TM) will need to happen soon, I think.

cfriedt commented:
* rebased (again) to resolve conflicts
* @carlescufi @nashif @stephanosio - can someone please merge this? It really just had merge conflict after merge conflict, has been rebased,  re-approved about a dozen times over the last almost 4 months with practically zero code changes.

aescolar commented:
> * rebased (again) to resolve conflicts
> 
>     * @carlescufi @nashif @stephanosio - can someone please merge this? It really just had merge conflict after merge conflict, has been rebased,  re-approved about a dozen times over the last almost 4 months with practically zero code changes.

@cfriedt note that only admins can override the compliance check failure..

stephanosio reviewed (approved):
LGTM
